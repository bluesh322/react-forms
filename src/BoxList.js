import React, { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import {v4 as uuid} from 'uuid';

const BoxList = () => {
    const INITIAL_STATE = [
        {
            width: 100,
            height: 100,
            color: "black",
            id: uuid()
        }
    ]
  const [boxes, setBoxes] = useState(INITIAL_STATE);

  const addBox = (box) => {
      let newBox = {...box, id: uuid()}
      setBoxes(boxes => [...boxes, newBox]);
  };

  const removeBox = (id) => {
      setBoxes(boxes.filter(b => b.id !== id));
  };

  return (
    <div className="content">
      <div className="NewBoxForm">
        <NewBoxForm addBox={addBox}/>
      </div>

      <div className="BoxList">
        {boxes.map(({width, height, color, id}) => (
          <Box
            width={width}
            height={height}
            color={color}
            removeBox={removeBox}
            id={id}
            key={id}
          />
        ))}
      </div>
    </div>
  );
};

export default BoxList;
