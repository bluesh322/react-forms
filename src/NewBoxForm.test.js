import React from "react";
import {render, fireEvent, queryAllByTitle } from "@testing-library/react";
import BoxList from "./BoxList";

it("renders without crashing", () => {
    render(<BoxList/>);
});

it("matches snapshot", () => {
    const {asFragment} = render(<BoxList/>);
    expect(asFragment()).toMatchSnapshot();
});

it("Should add new item", () => {
    const { queryByText, getByLabelText} = render(<BoxList/>);
    const widthInput = getByLabelText("width");
    const heightInput = getByLabelText("height");
    const colorInput = getByLabelText("color");
    const btn = queryByText("Add Box");
    fireEvent.change(widthInput, {target: { value: '100'}})
    fireEvent.change(heightInput, {target: { value: '100'}})
    fireEvent.change(colorInput, {target: { value: 'blue'}})
    fireEvent.click(btn);
    
    //what would I do here to test a new Box has been created?
    //Is this sufficient?
    expect(document.querySelectorAll('div.Box')[1]).toBeInTheDocument();

});