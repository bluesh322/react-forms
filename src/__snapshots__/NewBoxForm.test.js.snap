// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="content"
  >
    <div
      class="NewBoxForm"
    >
      <form>
        <label
          for="width"
        >
          width
        </label>
        <input
          id="width"
          name="width"
          placeholder="width"
          type="text"
          value=""
        />
        <label
          for="height"
        >
          height
        </label>
        <input
          id="height"
          name="height"
          placeholder="height"
          type="text"
          value=""
        />
        <label
          for="color"
        >
          color
        </label>
        <input
          id="color"
          name="color"
          placeholder="color"
          type="text"
          value=""
        />
        <button>
          Add Box
        </button>
      </form>
    </div>
    <div
      class="BoxList"
    >
      <div
        class="BoxArea"
      >
        <div
          class="Box"
          style="width: 100px; height: 100px; background-color: black; margin: 1em; padding: 1em;"
        />
        <button>
          X
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
